proposals/NNNN-negative-availability.md

```diff
@@ -0,0 +1,174 @@
+if #available(iOS 13, *) {
+
+} else {
+  loadMainWindow()
+}
+```
+
+## Readability
+
+As you might notice, the current way to achieve negative availability checks is by working around the current `#available` keyword. Because the availability condition is not parsed as an expression, it cannot be negated with regular boolean operations (`!`/`== false`). The way instead is to make use of the `else` clause, but as unavailability checks are not interested at all in the positive portion of the check, doing so will leave behind an empty `if` branch. For context, this problem does not exist in Objective-C as `if (@available(iOS 13.0, *) == NO)` is a valid expression.
+
+With the exception of this very specific case, an empty statement is a sure sign that there's something wrong in the developer's code, and it is likely that every unavailability check like this had to include a comment to indicate that it was done on purpose due to the compiler's limitations. In some cases, it might even be necessary to add an exclusion rule to the project's linters as most would assume that this is a mistake and incorrectly suggest that it can be fixed by negating the statement.
+
+This workaround has a clear negative impact on the readability of the unavailability check, as no one would expect an empty `if` statement to not be a coding mistake. Most developers will attempt to hide the problem by putting the positive portion in a single line:
+
+```swift
+// if NOT in iOS 13, load the window.
+// Post iOS 13 the window is loaded later in the lifecycle, in the SceneDelegate.
+if #available(iOS 13, *) {} else {
+  loadMainWindow()
+}
+```
+
+Unfortunately, this makes it easy to mistake it for a regular availability check. If the comment isn't clear that the statement is wrong on purpose, problems involving this check could easily land unnoticed in code-reviews. A less noisy way would be to put it under a `guard`:
+
+```swift
+guard #available(iOS 13, *) else {
+  loadMainWindow()
+  return
+}
+// no-op
+```
+
+However, this goes against the code style recommendations involving the usage of `guard`. The guarded part should be the happy path, which is not the case when doing unavailability checks. As shown, it's currently impossible to write an unavailability check that properly fits a developer's engineering expectations and Swift's general style guide. 
+
+Currently, any iOS application that supports UIScenes will face this issue and have to write this workaround. To describe it in a generic way, this issue will be encountered when dealing with any API addition or change that requires more than one availability condition to be implemented.
+
+## Usage of deprecated APIs
+
+A negative availability condition might also be necessary in cases where an API is marked as unavailable. While newer iOS versions will have migrated to new APIs, it's often the case that old versions must still respect the old requirements.
```

Comment:
Huh that's interesting. I was under the impression that `obsolete` would allow you to use the API in older versions but that's not the case.

```swift
@available(iOS, obsoleted: 14.0)
final class Obsolete {
}

if #available(iOS 14.0, *) {
    Obsolete() // fails
} else {
    Obsolete() // also fails!
}
```

Reply:
That example should work, if you set the iOS Deployment Target to 13.x or earlier.

Reply:
Actually, I can't find any examples where `obsoleted` is used in the iOS SDK. Maybe it's not supported in Objective-C headers, and the Swift-only frameworks are too recent to have obsoleted anything yet.